docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,216 @@
+- Feature Name: Partition Hibernation
+- Status: draft
+- Start Date: 2022-04-08
+- Authors: John Spray
+- Issue: (one or more # from the issue tracker)
+
+# Executive Summary
+
+Avoid idle partitions using system resources.
+
+## What is being proposed
+
+A new state for partitions, where they exist in the topics table
+but do not have active raft groups.
+
+## Why (short reason)
+
+To enable higher scale when a significant proportion of topics
+are idle.
+
+## How (short plan)
+
+Topics will be automatically placed into the hibernate state when
+they have been idle for more than a configurable threshold.
+
+## Impact
+
+Reduced cost of delivering free/cheap public-facing services, where
+users are likely to create topics for occasional or one-off use.
+
+# Motivation
+
+## Why are we doing this?
+
+Partitions have some runtime resource overhead, even if they aren't being
+used, and haven't been used for some time.
+
+To support public-facing Kafka services on which many users might create
+and then abandon idle topics, we could implement a "hibernate" mode
+for partitions that haven't been used for a long time.
+
+## What use cases does it support?
+
+Implementing a low-cost/no-cost public service, in which developers are likely
+to create unused topics with which to experiment.
+
+## What is the expected outcome?
+
+Systems will run smoothly with a higher number of partitions than
+they could before, as long as some of those partitions are idle.
+
+# Guide-level explanation
+
+## Introducing new named concepts.
+
+`Hibernated topic`: a topic whose partitions do not have active raft
+groups running in the cluster.  It cannot serve user I/O (aside from
+metadata requests) until it is woken up.
+
+`Waking a topic`: a control plane operation to change the status of
+the topic, followed by raft elections for all partitions within the
+topic.
+
+These concepts are not visible to applications.  They are optionally visible
+to system administrators optionally:
+- system monitoring telemetry, where we would surface indicators for how many partitions
+  and topics are currently hibernated.
+- tunable configuration properties, where they could control the delay til hibernation
+  if needed.
+
+
+# Reference-level explanation
+
+## High level design
+
+The topic table, which stores the definition of the topics & partitions
```

Comment:
That's an excellent point: this will generate a large number of messages at scale, although controllable somewhat by our choice of idle timeout.  If the timeout is like 24 hours, then a typical topic lifecycle for free tier eval would be something like one create, one hibernate, then eventually one delete.  The pathological case would be where they use the topic almost exactly at the same periodicity as the idle period.

I wonder if we will have some similar issues with the continuous data rebalancing, if we are regularly triggering large numbers of partition movements on a system that is running close to disk-full?

Reply:
in case of partition movment we need a persistence across node restarts, we do not really have other choice, we need two layers to move partitions 1) controller - to initiate move, create partition, and then eventually remove old replica and 2) raft layer to actually replicate data. In case of hibernation would that be sufficient for the partition to decide on its own if it needs hibernation. This way leader could send special RFC to followers and initiate hibernation process

Reply:
A decentralized version is interesting to think about... I guess it would mean each member of the raft group persisting something to disk that marked the partition hibernated (but not just a log entry, because then we'd still have to recover the log on startup even if hibernated).  Then to wake the partition, we'd RPC to each group member and have them remove the hibernate marker before starting.  If a follower was offline at this time, the leader would have to wake them up later when it saw them go online in the cluster health map.

If we ever wanted to know centrally which partitions were hibernated, we'd need to include that status in node health reports (as opposed to simply excluding hibernated partitions from the health reports in the original proposal).  In general it's not something we need to know centrally (aside from for monitoring purposes), but we might want it in the special case of moving topics between clusters + preferring to migrate those topics that are idle.

:thinking: